examples/recommendations/detect_and_apply_recommendations.rb

```diff
@@ -32,49 +36,90 @@ def detect_and_apply_recommendations(customer_id)
-  NUMBER_OF_RUNS.times do |i|
-    search_response = google_ads_service.search(
-      customer_id: customer_id,
-      query: query,
-    )
-    operations = search_response.map do |row|
-      client.operation.apply_recommendation do |aro|
-        aro.resource_name = row.recommendation.resource_name
-      end
-    end
 
-    unless operations.empty?
-      response = recommendation_service.apply_recommendation(
-        customer_id: customer_id,
-        operations: operations,
-      )
+  response = google_ads_service.search(
+    customer_id: customer_id,
+    query: query,
+  )
 
-      response.each do |result|
-        puts "Applied recommendation with resource name #{result.resource_name}."
-      end
-    end
+  operations = response.each do |row|
+    recommendation = row.recommendation
+
+    puts "Keyword recommendation ('#{recommendation.resource_name}') was found for "\
+      "campaign '#{recommendation.campaign}'."
 
-    if i < NUMBER_OF_RUNS - 1
-      puts "Waiting #{PERIOD_IN_SECONDS} seconds before applying more recommendations."
-      sleep(PERIOD_IN_SECONDS)
+    if recommendation.keyword_recommendation
+      keyword = recommendation.keyword_recommendation.keyword
+      puts "\tKeyword = '#{keyword.text}'\n\ttype = '#{keyword.match_type}'"
     end
+
+    build_recommendation_operation(client, recommendation.resource_name)
+  end
+  # [END detect_keyword_recommendations]
+
+  if operations
+    apply_recommendations(client, customer_id, recommendation.resource_name)
   end
 end
 
-if __FILE__ == $0
-  MAX_RESULT_SIZE = 2
-  NUMBER_OF_RUNS = 3
-  PERIOD_IN_SECONDS = 5
 
+# [START build_apply_recommendation_operation]
+def build_recommendation_operation(client, recommendation)
+  # If you have a recommendation_id instead of the resournce_name
+  # you can create a resource name from it like this:
+  # recommendation_resource =
+  #    client.path.recommendation(customer_id, recommendation_id)
+
+  operations = client.operation.apply_recommendation
+  operations.resource_name = recommendation_resource
+
+  # Each recommendation type has optional parameters to override the recommended
+  # values. This is an example to override a recommended ad when a
+  # TextAdRecommendation is applied.
+  # For details, please read
+  # https://developers.google.com/google-ads/api/reference/rpc/google.ads.google_ads.v1.services#google.ads.google_ads.v1.services.ApplyRecommendationOperation
+  #
+  # text_ad_parameters = client.resource.text_ad_parameters do |tap|
+  #   tap.ad = client.resource.ad do |ad|
+  #     ad.id = "INSERT_AD_ID_AS_INTEGER_HERE"
+  #   end
+  # end
+  # operation.text_ad = text_ad_parameters
+  #
+  # For more details, see:
+  # https://developers.google.com/google-ads/api/reference/rpc/latest/ApplyRecommendationOperation#apply_parameters
+
+  return operation
+end
+# [END build_apply_recommendation_operation]
+
+
+# [START apply_recommendation]
+def apply_recommendations(client, customer_id, operations)
```

Comment:
I think the other libs just include this logic inside the main function instead of pulling it into an apply_recommendations fn. Any reason not to do that here as well?

Reply:
I copied this from the python example. I can move it to the main function if most libraries have it there.

Reply:
They do but I guess it doesn't matter a whole lot, this is fine